### PR TITLE
Ruby: Cleanup flow through `self`

### PR DESCRIPTION
**Pull request by hvitved**
Since @hmac's https://github.com/github/codeql-ruby/pull/335, `self` is handled in the SSA library as any other variable (except we were missing support for captured variables, which required using `SsaImplSpecific::variableWrite` instead of `VariableWriteAccess` in `hasVariableWriteWithCapturedRead`). Consequently, we no longer need explicit flow from `self` parameters to all accesses, as this is now covered by normal parameter-to-first-use flow, followed by adjacent-use-use flow. 

The last two commits are unrelated performance tweaks (though the performance problem fixed by the latter commit was amplified by letting the SSA library handle flow through captured `self` variables).

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/ast/internal/Variable.qll [hunk 1]
```diff
@@ -366,7 +366,23 @@ private module Cached {
...
+      if access instanceof SelfVariableAccess
+      then
+        // ```
+        // class C
+        //   def self.m // not a captured access
+        //   end
+        // end
+        // ```
+        not scope2 instanceof Toplevel or
+        not access = any(SingletonMethod m).getObject()
```
Comment (alexrford) [inline ruby/ql/lib/codeql/ruby/ast/internal/Variable.qll:383]: I personally find this a bit easier to read as:

```suggestion
        not (scope2 instanceof Toplevel and access = any(SingletonMethod m).getObject())
```

**Review by alexrford — COMMENTED** [review]
Looks good - I'm not particularly familiar with the code, so this review is a bit superficial.

**Review by alexrford — DISMISSED** [review]

File: ruby/ql/lib/codeql/ruby/ast/internal/Variable.qll [hunk 2]
```diff
@@ -366,7 +366,23 @@ private module Cached {
...
+      else any()
+    )
```
Comment (hmac) [inline ruby/ql/lib/codeql/ruby/ast/internal/Variable.qll:384]: I don't think it was introduce by this change, but I've just noticed that reads of variables captured by a lambda get interpreted as methods calls with an implicit `self`, which seems wrong:
```rb
x = 1
y = -> { x } # x is interpreted as self.x()
```

Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/ast/internal/Variable.qll:384]: Let's fix this follow-up.

**Review by hmac — COMMENTED** [review]
The `self` changes look good to me. The the third commit, 722b1e11d11fb741508066f5d95c2357db20e2df, seems unrelated to the other changes. I'm also struggling to disentangle what is new code and what is just refactored code. Is it possible to add a test case that shows the change in behaviour? Or even a snippet of code in a doc comment. That would make a lot easier to know what is going on.

**Comment by hvitved** [comment]
> The `self` changes look good to me. The the third commit, [722b1e1](https://github.com/github/codeql/commit/722b1e11d11fb741508066f5d95c2357db20e2df), seems unrelated to the other changes.

While it is not strictly related, it implements a performance improvement that was required to avoid slow-down following the change to `self` flow. Otherwise, there would have been, on some projects, huge (false-positive) flow through captured `self` variables.

> I'm also struggling to disentangle what is new code and what is just refactored code.

The first commit is the only commit that changes `self` flow. 

> Is it possible to add a test case that shows the change in behaviour? Or even a snippet of code in a doc comment. That would make a lot easier to know what is going on.

I did not add any tests, since I thought it was already covered sufficiently. For example, [the changes to `DataflowStep.expected`](https://github.com/github/codeql/pull/7084/files?show-viewed-files=true&file-filters%5B%5D=#diff-28b9ac94b68067287db2d3ec8dc1a154ff7b57897f416b47d035c8a37ec025f2L74) show that we no longer have flow from implicit `self` parameters to all accesses (instead, we have flow only to the first access(es), and from there adjacent use-use flow, like for normal variables).

**Review by hmac — APPROVED** [review]

**Pull request merged**